Backfilling fails for SCIP when solving hits limit

E.g. with a low `stallnodelimit`, this error is thrown:
```
    cvxpy_translation.backfill_problem(prob, model)
.venv/lib/python3.12/site-packages/cvxpy_translation/interface.py:93: in backfill_problem
    backfill_scip_problem(problem, model, compilation_time, solve_time)
.venv/lib/python3.12/site-packages/cvxpy_translation/scip/interface.py:100: in backfill_problem
    problem.unpack(solution)
.venv/lib/python3.12/site-packages/cvxpy/problems/problem.py:1510: in unpack
    raise ValueError("Cannot unpack invalid solution: %s" % solution)
E   ValueError: Cannot unpack invalid solution: Solution(status=solver_error, opt_val=None, primal_vars={}, dual_vars={}, attr={'solver_specific_stats': <pyscipopt.scip.Model object at 0x727d1a55a7a0>, 'solve_time': 0.190625, 'num_iters': 1})
```
This is due to the fact that the status gets mapped to `solver_error` from CVXPY which cannot be handled by `problem.unpack`. So this might be a much more general problem that occurs whenever solving fails.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,10 @@
 
 - Fix translation of constant objective terms for both SCIP and Gurobi
   ([#197](https://github.com/jonathanberthias/cvxpy-translation/pull/197))
+- Fix backfilling for SCIP when presolve removes constraints
+  ([#201](https://github.com/jonathanberthias/cvxpy-translation/pull/201))
+- Fix backfilling for SCIP when a user limit is hit before any solution is found
+  ([#202](https://github.com/jonathanberthias/cvxpy-translation/pull/202))
 
 ## [2.0.0] - 2025-08-23
 

--- a/src/cvxpy_translation/scip/interface.py
+++ b/src/cvxpy_translation/scip/interface.py
@@ -6,7 +6,6 @@ from typing import Union
 
 import cvxpy as cp
 import cvxpy.settings as cp_settings
-import cvxpy.settings as s
 import numpy as np
 import numpy.typing as npt
 import pyscipopt as scip
@@ -102,7 +101,7 @@ def backfill_problem(
     # so we lie about the status to be sure the processing is still done
     original_status = solution.status
     if model.getNSols() == 0:
-        solution.status = s.INFEASIBLE
+        solution.status = cp_settings.INFEASIBLE
 
     problem.unpack(solution)
     solution.status = original_status
@@ -137,7 +136,11 @@ def extract_solution_from_model(model: scip.Model, problem: cp.Problem) -> Solut
     if model.getNSols() == 0:
         # Make status more accurate - user limit statuses are like:
         # timelimit, nodelimit, bestsollimit, etc.
-        status = s.USER_LIMIT if model.getStatus().endswith("limit") else s.SOLVER_ERROR
+        status = (
+            cp_settings.USER_LIMIT
+            if model.getStatus().endswith("limit")
+            else cp_settings.SOLVER_ERROR
+        )
         if CVXPY_VERSION >= (1, 2, 0):
             # attr was added in https://github.com/cvxpy/cvxpy/pull/1270
             return failure_solution(status, attr)

--- a/src/cvxpy_translation/scip/interface.py
+++ b/src/cvxpy_translation/scip/interface.py
@@ -130,17 +130,6 @@ class UnavailableDualError(ValueError):
     pass
 
 
-LIMIT_STATUSES = {
-    "timelimit",
-    "nodelimit",
-    "totalnodelimit",
-    "stallnodelimit",
-    "memlimit",
-    "sollimit",
-    "bestsollimit",
-}
-
-
 def extract_solution_from_model(model: scip.Model, problem: cp.Problem) -> Solution:
     attr = {
         cp_settings.EXTRA_STATS: model,
@@ -149,8 +138,9 @@ def extract_solution_from_model(model: scip.Model, problem: cp.Problem) -> Solut
     }
     status = scip_conif.STATUS_MAP[model.getStatus()]
     if model.getNSols() == 0:
-        # Make status more accurate
-        status = s.USER_LIMIT if model.getStatus() in LIMIT_STATUSES else s.SOLVER_ERROR
+        # Make status more accurate - user limit statuses are like:
+        # timelimit, nodelimit, bestsollimit, etc.
+        status = s.USER_LIMIT if model.getStatus().endswith("limit") else s.SOLVER_ERROR
         if CVXPY_VERSION >= (1, 2, 0):
             # attr was added in https://github.com/cvxpy/cvxpy/pull/1270
             return failure_solution(status, attr)

--- a/src/cvxpy_translation/scip/interface.py
+++ b/src/cvxpy_translation/scip/interface.py
@@ -89,9 +89,6 @@ def set_params(model: scip.Model, params: ParamDict) -> None:
         model.setParam(key, param)
 
 
-CP_STATUS_WITH_SOLUTION = {cp.OPTIMAL, cp.OPTIMAL_INACCURATE}
-
-
 def backfill_problem(
     problem: cp.Problem,
     model: scip.Model,

--- a/src/cvxpy_translation/scip/interface.py
+++ b/src/cvxpy_translation/scip/interface.py
@@ -6,6 +6,7 @@ from typing import Union
 
 import cvxpy as cp
 import cvxpy.settings as cp_settings
+import cvxpy.settings as s
 import numpy as np
 import numpy.typing as npt
 import pyscipopt as scip
@@ -15,7 +16,6 @@ from cvxpy.problems.problem import SolverStats
 from cvxpy.reductions.solution import Solution
 from cvxpy.reductions.solution import failure_solution
 from cvxpy.reductions.solvers.conic_solvers import scip_conif
-from cvxpy.settings import SOLUTION_PRESENT
 
 from cvxpy_translation import CVXPY_VERSION
 from cvxpy_translation.scip.translation import Translater
@@ -89,6 +89,9 @@ def set_params(model: scip.Model, params: ParamDict) -> None:
         model.setParam(key, param)
 
 
+CP_STATUS_WITH_SOLUTION = {cp.OPTIMAL, cp.OPTIMAL_INACCURATE}
+
+
 def backfill_problem(
     problem: cp.Problem,
     model: scip.Model,
@@ -97,7 +100,17 @@ def backfill_problem(
 ) -> None:
     """Update the CVXPY problem with the solution from the SCIP model."""
     solution = extract_solution_from_model(model, problem)
+
+    # problem.unpack does not handle statuses other than optimal and inf_or_unbd
+    # so we lie about the status to be sure the processing is still done
+    original_status = solution.status
+    if model.getNSols() == 0:
+        solution.status = s.INFEASIBLE
+
     problem.unpack(solution)
+    solution.status = original_status
+    # overwrite the status
+    problem._status = original_status  # noqa: SLF001
 
     if CVXPY_VERSION >= (1, 4):
         # class construction changed in https://github.com/cvxpy/cvxpy/pull/2141
@@ -117,6 +130,17 @@ class UnavailableDualError(ValueError):
     pass
 
 
+LIMIT_STATUSES = {
+    "timelimit",
+    "nodelimit",
+    "totalnodelimit",
+    "stallnodelimit",
+    "memlimit",
+    "sollimit",
+    "bestsollimit",
+}
+
+
 def extract_solution_from_model(model: scip.Model, problem: cp.Problem) -> Solution:
     attr = {
         cp_settings.EXTRA_STATS: model,
@@ -124,7 +148,9 @@ def extract_solution_from_model(model: scip.Model, problem: cp.Problem) -> Solut
         cp_settings.NUM_ITERS: model.lpiGetIterations(),
     }
     status = scip_conif.STATUS_MAP[model.getStatus()]
-    if status not in SOLUTION_PRESENT:
+    if model.getNSols() == 0:
+        # Make status more accurate
+        status = s.USER_LIMIT if model.getStatus() in LIMIT_STATUSES else s.SOLVER_ERROR
         if CVXPY_VERSION >= (1, 2, 0):
             # attr was added in https://github.com/cvxpy/cvxpy/pull/1270
             return failure_solution(status, attr)

--- a/tests/test_backfill.py
+++ b/tests/test_backfill.py
@@ -1,4 +1,6 @@
 import cvxpy as cp
+import numpy as np
+import pytest
 
 from cvxpy_translation import backfill_problem
 from cvxpy_translation import build_model
@@ -21,3 +23,30 @@ def test_empty_presolved_model_grb():
     model = build_model(problem, cp.GUROBI)
     model.optimize()
     backfill_problem(problem, model)
+
+
+@pytest.mark.parametrize(
+    ("param", "value", "status"),
+    [
+        ("limits/time", 0, "timelimit"),
+        ("limits/nodes", 0, "nodelimit"),
+        ("limits/totalnodes", 0, "totalnodelimit"),
+        ("limits/stallnodes", 0, "stallnodelimit"),
+        ("limits/memory", 0, "memlimit"),
+        ("limits/solutions", 0, "sollimit"),
+        ("limits/bestsol", 0, "bestsollimit"),
+    ],
+)
+def test_user_limit_scip(param: str, value: int, status: str):
+    """If the user limit is reached, the problem should still be backfilled."""
+    x = cp.Variable(50, boolean=True)
+    # some random non-trivial problem
+    problem = cp.Problem(
+        cp.Minimize(cp.sum(x**2 + cp.log1p(x))), [x @ np.arange(x.size) == 13]
+    )
+    model = build_model(problem, cp.SCIP)
+    model.setParam(param, value)
+    model.optimize()
+    assert model.getStatus() == status
+    backfill_problem(problem, model)
+    assert problem.status == cp.USER_LIMIT

